pkg/sources/adapter/awssnssource/probe/probe.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package probe contains facilities for asserting the readiness of a
+// multi-tenant receive adapter.
+package probe
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/triggermesh/triggermesh/pkg/apis/sources/v1alpha1"
+	listersv1alpha1 "github.com/triggermesh/triggermesh/pkg/client/generated/listers/sources/v1alpha1"
+	"github.com/triggermesh/triggermesh/pkg/sources/adapter/common/router"
+)
+
+// ReadinessChecker can assert the readiness of a component.
+type ReadinessChecker interface {
+	IsReady() (bool, error)
+}
+
+// AdapterReadyChecker asserts the readiness of a receive adapter.
+type AdapterReadyChecker struct {
+	srcLister listersv1alpha1.AWSSNSSourceLister
+
+	srcRouter *router.Router
+
+	sync.RWMutex
+	isReady bool
+}
+
+var _ ReadinessChecker = (*AdapterReadyChecker)(nil)
+
+// NewAdapterReadyChecker returns an AdapterReadyChecker initialized with the
+// given Lister and Router.
+func NewAdapterReadyChecker(ls listersv1alpha1.AWSSNSSourceLister, r *router.Router) *AdapterReadyChecker {
+	return &AdapterReadyChecker{
+		srcLister: ls,
+		srcRouter: r,
+	}
+}
+
+// IsReady implements ReadinessChecker.
+// It checks whether the adapter has registered a handler for all observed sources.
+//
+// NOTE(antoineco): we might want to revisit this in the future, because the
+// following implementation details are currently leaking into this
+// ReadinessChecker implementation:
+//
+// - The fact that the adapter exposes a health endpoint via the HTTP router.
+// - The exact URL path of this health endpoint.
+// - The fact that a handler isn't registered for sources which don't have a valid sink.
+//
+// We consider this tight coupling acceptable for the time being, because this
+// is a sub-package of the main adapter package for the SNS source
+// specifically.
```

Comment:
After sending the PR, I realized that the coupling between the adapter and the checker are so tight that it might make no sense at all to have the checker inside a separate package.

Comments/concerns welcome.